content/docs/pipelines/reference/component-spec.md

```diff
@@ -0,0 +1,135 @@
+
+* **Metadata:** Name, description, and other metadata.
+* **Interface (inputs and outputs):** Name, type, default value.
+* **Implementation:** How to run the component, given the input arguments.
+
+## Example of a component specification
+
+A component specification takes the form of a YAML file, `component.yaml`. Below
+is an example:
+
+```
+name: xgboost4j - Train classifier
+description: Trains a boosted tree ensemble classifier using xgboost4j
+
+inputs:
+- {name: Training data URI}
+- {name: Rounds, type: Integer, default: '30', help: Number of training rounds}
+
+outputs:
+- {name: Trained model, type: XGBoost model, help: Trained XGBoost model}
+
+implementation:
+  container:
+    image: gcr.io/ml-pipeline/xgboost-classifier-train@sha256:b3a64d57
+    command: [
+      /ml/train.py,
+      --train-uri, {inputValue: Training data URI},
+      --rounds,    {inputValue: Rounds},
+      --out-model, {outputPath: Trained model},
+    ]
+```
+
+See some examples of real-world 
+[component specifications](https://github.com/kubeflow/pipelines/search?q=filename%3Acomponent.yaml&unscoped_q=filename%3Acomponent.yaml).
+
+## Detailed specification (ComponentSpec)
+
+This section describes the 
+[ComponentSpec](https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/components/_structures.py).
+
+### Metadata
+
+* `name`: Human-readable name of the component.
+* `description`: Description of the component.
+* `metadata`: Standard object's metadata:
+
+    * `annotations`: An unstructured key value map
+        stored with a resource. External tools can set this property to store and
+        retrieve arbitrary metadata. Annotations are not queryable and should be
+        preserved when modifying objects. See more information in the
+        [Kubernetes user guide](http://kubernetes.io/docs/user-guide/annotations).
+
+    * `labels`: A map of string keys and values that can be used to
+        organize and categorize (scope and select) objects. May match selectors
+        of replication controllers and services. See more information in the
+        [Kubernetes user guide](http://kubernetes.io/docs/user-guide/labels).
+
+### Interface
+
+* `inputs` and `outputs`:
+    Specifies the list of inputs/outputs and their properties. Each input or
+    output has the following properties:
+
+    * `name`: Human-readable name of the input/output. Name must be
+        unique inside the inputs or outputs section, but an output may have the
+        same name as an input.
+    * `description`: Human-readable description of the input/output.
+    * `default`: Specifies the default value for an input. Only
+        valid for inputs.
+    * `type`: Specifies the type of input/output. The types are used
+        as hints for pipeline authors and can be used by the pipeline system/UI
+        to validate arguments and connections between components. Basic types
+        are **String**, **Integer**, **Number**, **Boolean**.
```

Comment:
Is this a full list of supported types? If not - can we have a reference here to the full list?

Reply:
Done.